packages/devextreme-react/src/core/component-base-func.tsx

```diff
@@ -0,0 +1,397 @@
+        updateCssClasses,
+        setInlineStyles,
+        props,
+      ]);
+
+      const onComponentUnmounted = useCallback(() => {
+        removalLocker?.lock();
+
+        if (instance.current) {
+          const dxRemoveArgs: DXRemoveCustomArgs = { isUnmounting: true };
+
+          childNodes.current?.forEach((child) => child.parentNode?.removeChild(child));
+
+          if (element.current) {
+            events.triggerHandler(element.current, DX_REMOVE_EVENT, dxRemoveArgs);
+          }
+
+          instance.current.dispose();
+        }
+        optionsManager.current.dispose();
+
+        removalLocker?.unlock();
+      }, [
+        removalLocker,
+        instance.current,
+        childNodes.current,
+        element.current,
+        optionsManager.current,
+      ]);
+
+      useEffect(() => {
+        onComponentMounted();
+
+        return () => {
+          onComponentUnmounted();
+        };
+      }, []);
+
+      useEffect(() => {
+        onComponentUpdated();
+      });
+
+      useImperativeHandle(ref, () => (
+        {
+          getInstance() {
+            return instance.current;
+          },
+          getElement() {
+            return element.current;
+          },
+          createWidget(el) {
+            createWidget(el);
+          },
+        }
+      ));
+
+      const _renderChildren = useCallback(() => {
+        if (renderChildren) {
+          return renderChildren();
+        }
+
+        // @ts-expect-error TS2339
+        const { children } = props;
+        return children;
+      }, [props, renderChildren]);
+
+      const renderPortal = useCallback(() => portalContainer.current && createPortal(
+        _renderChildren(),
+        portalContainer.current,
+      ), [portalContainer.current, _renderChildren]);
+
+      const renderContent = useCallback(() => {
+        // @ts-expect-error TS2339
+        const { children } = props;
+
+        return isPortalComponent && children
+          ? React.createElement('div', {
+            ref: (node: HTMLDivElement | null) => {
+              if (node && portalContainer.current !== node) {
+                portalContainer.current = node;
+                setForceUpdateToken(Math.random());
+              }
+            },
+            style: { display: 'contents' },
+          })
+          : _renderChildren();
+      }, [
+        props,
+        isPortalComponent,
+        portalContainer.current,
+        _renderChildren,
+      ]);
+
+      return React.createElement(
+        React.Fragment,
+        {},
+        React.createElement(
+          'div',
+          getElementProps(),
+          renderContent(),
+          React.createElement(TemplateManager, {
+            init: setTemplateManagerHooks,
+          }),
+        ),
+        isPortalComponent && renderPortal(),
+      );
+    },
+  ),
+) as <P extends IHtmlOptions>(props: P & ComponentBaseProps, ref: React.ForwardedRef<ComponentBaseRef>) => ReactElement | null;
```

Comment:
Note for reviewers: This type casting is required because TypeScript is unable to infer types correctly if a functional component is a generic function and at the same time uses HOC functions (like `memo` or `forwardRef`).